pkg/security/probe/custom_events.go

```diff
@@ -0,0 +1,218 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build linux
+
+package probe
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/security/rules"
+	"github.com/DataDog/datadog-agent/pkg/security/secl/eval"
+)
+
+const (
+	// LostEventsRuleID is the rule ID for the lost_events_* events
+	LostEventsRuleID = "lost_events"
+	// RuleSetLoadedRuleID is the rule ID for the ruleset_loaded events
+	RuleSetLoadedRuleID = "ruleset_loaded"
+	// NoisyProcessRuleID is the rule ID for the noisy_process events
+	NoisyProcessRuleID = "noisy_process"
+	// AbnormalPathRuleID is the rule ID for the abnormal_path events
+	AbnormalPathRuleID = "abnormal_path"
+	// ForkBombRuleID is the rule ID for the fork_bomb events
+	ForkBombRuleID = "fork_bomb"
+)
+
+// AllCustomRuleIDs returns the list of custom rule IDs
+func AllCustomRuleIDs() []string {
+	return []string{
+		LostEventsRuleID,
+		RuleSetLoadedRuleID,
+		NoisyProcessRuleID,
+		AbnormalPathRuleID,
+		ForkBombRuleID,
+	}
+}
+
+// CustomEvent is used to send custom security events to Datadog
+type CustomEvent struct {
+	eventType   string
+	tags        []string
+	marshalFunc func() ([]byte, error)
+}
+
+// Clone returns a copy of the current CustomEvent
+func (ce *CustomEvent) Clone() CustomEvent {
+	return CustomEvent{
+		eventType:   ce.eventType,
+		tags:        ce.tags,
+		marshalFunc: ce.marshalFunc,
+	}
+}
+
+// GetTags returns the tags of the custom event
+func (ce *CustomEvent) GetTags() []string {
+	return append(ce.tags, "type:"+ce.GetType())
+}
+
+// GetType returns the type of the custom event
+func (ce *CustomEvent) GetType() string {
+	return ce.eventType
+}
+
+// MarshalJSON is the JSON marshaller function of the custom event
+func (ce *CustomEvent) MarshalJSON() ([]byte, error) {
+	if ce.marshalFunc != nil {
+		return ce.marshalFunc()
+	}
+	return nil, nil
+}
+
+// String returns the string representation of a custom event
+func (ce *CustomEvent) String() string {
+	d, err := json.Marshal(ce)
+	if err != nil {
+		return err.Error()
+	}
+	return string(d)
+}
+
+func fillCustomRule(ruleDef *rules.RuleDefinition, event *CustomEvent) (*rules.Rule, *CustomEvent) {
+	return &rules.Rule{
+		Rule:       &eval.Rule{ID: ruleDef.ID},
+		Definition: ruleDef,
+	}, event
+}
+
+// NewEventLostReadEvent returns the rule and a populated custom event for a lost_events_read event
+func NewEventLostReadEvent(mapName string, perCPU map[int]int64, timestamp time.Time) (*rules.Rule, *CustomEvent) {
+	return fillCustomRule(&rules.RuleDefinition{
+		ID: LostEventsRuleID,
+	}, &CustomEvent{
+		eventType: "lost_events_read",
+		marshalFunc: func() ([]byte, error) {
+			return json.Marshal(struct {
```

Comment:
What about creating a dedicated struct for this and generate fast marshallers using easyjson ?

Reply:
Done 👍🏻